### DISCUSSION TITLE
More games

### DISCUSSION THREAD

**Discussion by harbaum** [harbaum/galagino#16] [Ideas]
Galagino can support some more games. But there are limitations:

* The emulation performance is limited to machines of the early 80s. Galagas three z80's are already close to the limit
* the screen needs to be upright with at most 240 x 320 pixels
* regular joystick controls have to be sufficient

One game that fulfils all of these is Frogger.

**Comment by C64fan** [comment]
Nice one's would be Galaxian and 1942.

**Reply by harbaum** [reply]
Galaxian is very similar to galaga. Having both of them in the same setup might seem somewhat redundant.

1942 is indeed a cool game. I need to look up the hardware specs. Also it needs an additional button for looping which may perhaps be combined/shared with the start button.

Another possible game is digdug.

**Reply by C64fan** [reply]
I aggree, instead of Galaxian it is preferred Phoenix which has nice levels with many different birds.

**Comment by Pep00000001** [comment]
Tron possible?
2xZ80 2xAY 8910

**Reply by harbaum** [reply]
This page claims that the Tron arcade has a 512x480 video resolution which is more than the 320*240 the galagino display supports.

http://www.arcaderestoration.com/games/9774/Tron.aspx

**Comment by harbaum** [comment]
I just added Frogger

**Comment by harbaum** [comment]
1942 seems to fit the bill as well. It uses two AY-8910 audio chips. That's the same chip frogger uses for audio and which has partially (only the functions frogger uses) been added already.

**Comment by lemonfresh33** [comment]
Pooyan might be doable- seems to be z80s and ay8910.  

Ref. https://github.com/opengateware/arcade-pooyan

How does one go about "porting" Mame ROMs into this emulator?  Can I help?

**Reply by harbaum** [reply]
You don't port the ROMs. The ROMs are the only part of the aracde machine that can be copied. But the hardware itself has to be re-written. That's the main thing an emulator does: To implement the hardware of the original machine  in software. If you want to port another machine, then you need to understand the hardware of that machine and re-write it.

**Reply by lemonfresh33** [reply]
Yeah, sorry, I get that...

By "port", I was referring to exactly that process.  I have found  the MAME "driver" (probably also the wrong term, I have found the code in Mame that "implements the hardware" for Pooyan and Frogger).  ( I was referring to that when I was using the term "port", not specifically the ROM data).  

I understand the Frogger and Pooyan hardware are very similar, so I'm thinking about taking your existing Frogger code and fiddling with it, by comparing the differences in the Mame code which would presumably be "similar" to each other, e.g. perhaps with the ROMS at different memory addresses, different locations for I/O ports etc.

Do you have any tips before I dive in and have a go?

**Reply by harbaum** [reply]
There's no rule how to do it. But here's how I do it:

I start by just adding the ROMs to the CPU and adding a few printf's that catch memory accesses to parts I haven't implemented yet. That will likely trigger some RAM write immediately. So I add RAM to that area. Then the code will run further and will start to read and write IO areas. Step by step I add these as well. At one point the CPU will start setting up graphics in memory. That's when I implement the graphics rendering. And then I start to see things. Probably sprites at the wrong location with wrong colors and the like ... these are then fixed step-by-step.

I am doing this on a Linux PC and for simplicity I write some kind of PC emulator for Galagino. That wouldn't help me with performance issues as the PC is way faster than the ESP32. But being able to run the majority of the code on the PC itself eases debugging of the logic.

**Comment by jf1452** [comment]
Awesome work harbaum, some of my favorites, from the 80's already, you have two listed:
Mr Do!
Star Force
Space Invaders
Asteroids (might be a bit blocky on a low res screen)

**Comment by lemonfresh33** [comment]
Most excellent.  Now I just need to find some spare time. Always the
hardest part!

Most excellent project by the way!

**Comment by harbaum** [comment]
Digdug has been added.

**Comment by harbaum** [comment]
1942 has been added. This actually affects the other implementations as well and after 1942 was added digdug needed some more optimizations to run with the necessary performance. Interestingly, Didug is still the game in this setup that has the highest performance demands.

1942 on the other hand needs less CPU power (4Mhz + 3Mhz Z80 CPUs vs e.g. 3 * 3MHz Z80 in Digdug). But it comes with the biggest ROMs. And since I offload some CPU intense stuff and e.g. pre-rotate tiles and icons, the flash requirements of some games are much higher than they were on the original machines. The background tiles in 1942 are 48 kilobytes in the original ROM. Pre-rotated and expanded to 4 bits per pixel, the same tiles occupy 256kBytes in galagino.

So in total all six games now sum up to 1.25 Megabytes and 1.3 Megabytes is the limit of the default memory layout on a ESP32.

**Reply by jf1452** [reply]
Excellent work @harbaum , the most incredible use of an ESP32 I have seen.

Now you have filed the memory, I hope this isn't the end of the road; do you plan to add more games for different configurations? I can imagine a mini arcade full of these wonderful cabinets.

**Reply by harbaum** [reply]
The flash isn't really full. It's just that the default config allows for 1.31MB at most. The flash is actually 4MB. But the user would have to explicitly select a different flash Layout in the Arduino IDE if more than 1.31MB are to be used.

So there may be more games to come ...

**Reply by witnessmenow** [reply]
"Huge App" time :)

**Comment by C64fan** [comment]
Happy with Galaga and 1942, many thanks.
Will update the Lilygo board.
Respect for all the work on this great project.

**Comment by Pete-IA** [comment]
Hello
I think you could add a lot of games, you just couldn't have all of them at once. It would be a matter of selecting them in the config.h until the flash was complete. It could even be done separately, so that one code = one game, and so you could do what jf1452 says, make a mini arcade with fully functional mini-reproductions. As I mentioned, I am making a scale replica of the Pacman machine which I will mount with a screen + a dedicated ESP32 for pacman game.
Although I am going little by little, I am not completely clear about the implementation of the games, if a guide were published to make the approximations and program the hardware emulation (how to carry out the first load, how to see the printf...) as Harbaum commented on June 15, maybe we could lend a hand to implement more games. 
Thank you very much for all the work and for sharing it.

**Comment by rhuwald** [comment]
Hello,

great work. Spent hours and hours with solving and building the arcade. Galaga is great, also 1942!

What about Bomb Jack from Thekan? It uses 2 Z8 (4 Mhz and 3.1 Mhz) and for sound 3 AY.3.8910...

best wishes from Muenster, DE
Ralf

**Comment by harbaum** [comment]
Bomb Jack actually sounds like a good candidate. I'll take a closer look at that.  Three sound chips ... wow. 1942 uses two of them and sounds like they wouldn't even need one at all ...

**Comment by ManicMiner-Es** [comment]
Dear friends ...

Is there any guide, documentation or clues about wich kind of file is every one contained in the zipped "roms"? As noobie with "roms" handling still lack of tech. information sources.

Any help will be appreciated with the promise to share any advance in this forum..

Regards

**Comment by VogonCaptain** [comment]
Excellent work on this project, thank you!  As for adding more games, one of my favorites is missile command... Is it possible?

**Comment by jestock1** [comment]
Awesome job on the emulator and thank you for the contribution, @harbaum!

I revised the code to add VGA output (only 8-bit, but good enough for me) and I am trying to add Ms. Pac-Man. I have successfully implemented the necessary decryption and have the ROM data set up in memory. I am able to receive the decode latch write to enable decoding and the game displays the test screen like Pac-Man does, but then it just displays a black screen. I've confirmed that the emulator is still running, but the game does not seem to proceed beyond that point. If I set the board test switch to on, it will perform the memory test and then remain on the memory test results screen.

I am getting reads and writes for all addresses < 0x4000, as well as 0x5000, 0x5040 and 0x5060 , but not for anything between 0x8000 and 0xA000. I've tried examining the memory and ROM data that is loaded into memory, but I'm lost on what to try next.

Do you have any suggestions/recommendations for further steps I can take to diagnose why it won't load into the game? Any advice or pointers you could provide would be greatly appreciated.

Thank you again for all your work!

**Comment by retronicdesign** [comment]
Any project to port Ms Pac-Mac? I know it's pretty close to Pac-Man with some exceptions. I find Ms Pac-Mac a bit more challenging than the original Pac-Man.

Great Project !! @harbaum

**Comment by harbaum** [comment]
I actually find "quite similar" games rather uninteresting. But IMHO it shouldn't be too much work to make Ms Pacman an option to add it instead if Pacman. But that's up to someone else ...

**Comment by ashep1337** [comment]
Tetris might be a good add if its possible.

Great work harbaum! I got things working on a CYD with a fully 3d printed case - its got a nice gameboy advanced feel to it

**Comment by harbaum** [comment]
~~Indeed, Tetris seems like a nice candidate. But I cannot find the full specs of the arcade machine online. The page https://www.arcade-museum.com/Videogame/tetris suggests that it's a single 6502 with dual pokey. That should be doable. But there's more info needed like how exactly the video worked etc~~

Actually, no. The Tetris arcade uses a horizontally mounted landscape screen. The Galagino uses a vertically mounted portrait screen.

**Comment by VirtualClaudioBoy** [comment]
Three years have passed, and I reread what I wrote with hindsight:
The project continued without Harbaum, and new games were ported to the Galagino, including some of those mentioned: Space Invaders, Ms Pacman, Mr. Do, Star force and Galaxian.

**Reply by retronicdesign** [reply]
OMG! Thank you so much for this! 🙏  I configured the code for my old CYD 2432S028 (very cheap yellow display, 5 bucks on Aliexpress nowadays). Works very well with my custom joypad using the pcf8574 i2c 8 bit IO extender.
<img width="2405" height="1771" alt="vcyd" src="https://github.com/user-attachments/assets/180921d3-2b9d-47e4-be4d-ea33fb7ea355" />
